drivers/Aqara/aqara-lock/src/init.lua

```diff
@@ -0,0 +1,233 @@
+    device:emit_event(TamperAlert.tamper.clear())
+  elseif evt_value == 4 then
+    device:emit_event(TamperAlert.tamper.detected())
+  end
+end
+
+local function event_battery_handler(driver, device, evt_name, evt_value)
+  device:emit_event(Battery.battery(evt_value))
+end
+
+local function event_temper_alert_handler(driver, device, evt_name, evt_value)
+  device:emit_event(TamperAlert.tamper.detected())
+end
+
+local resource_id = {
+  { id = "13.31.85", event_name = "locked",       event_handler = event_lock_handler },
+  { id = "13.17.85", event_name = "not fully locked", event_handler = event_door_handler },
+  { id = "13.48.85", event_name = "manual",       event_handler = event_unlock_handler },
+  { id = "13.51.85", event_name = "manual",       event_handler = event_unlock_handler },
+  { id = "13.42.85", event_name = "fingerprint",  event_handler = event_unlock_handler },
+  { id = "13.43.85", event_name = "keypad",       event_handler = event_unlock_handler },
+  { id = "13.44.85", event_name = "rfid",         event_handler = event_unlock_handler },
+  { id = "13.45.85", event_name = "bluetooth",    event_handler = event_unlock_handler },
+  { id = "13.90.85", event_name = "command",      event_handler = event_unlock_handler },
+  { id = "13.46.85", event_name = "keypad",       event_handler = event_unlock_handler },
+  { id = "13.56.85", event_name = "",             event_handler = event_battery_handler },
+  { id = "13.32.85", event_name = "",             event_handler = event_temper_alert_handler },
+}
+
+local function request_generate_shared_key(device)
+  device:send(cluster_base.write_manufacturer_specific_attribute(device,
+    PRI_CLU, PRI_ATTR, MFG_CODE, data_types.OctetString, "\x2B"))
+end
+
+local function lock_state_handler(driver, device, value, zb_rx)
+  local shared_key = device:get_field(SHARED_KEY)
+  local param = value.value
+  local command = string.sub(param, 0, 1)
+
+  if command == "\x3E" then
+    -- recv lock_pub_key
+    local locks_pub_key = string.sub(param, 2, string.len(param))
+    local mn_id = "Id3A"
+    local setup_id = "006"
+    local product_id = ""
+    local res, err = security.get_aqara_secret(device.zigbee_eui, locks_pub_key, "", mn_id, setup_id,
+      product_id)
+    if res then
+      print(res)
+    end
+  elseif shared_key == nil then
+    request_generate_shared_key(device)
+  elseif command == "\x93" then
+    local opts = { cipher = "aes256-ecb", padding = false }
+    local raw_key = base64.decode(shared_key)
+    local raw_data = string.sub(param, 2, string.len(param))
+    local msg = security.decrypt_bytes(raw_data, raw_key, opts)
+    local text = string.sub(msg, 5, string.len(msg))
+    local payload = string.sub(text, 4, string.len(text))
+    local func_id = toValue(payload, 1, 1) .. "." .. toValue(payload, 2, 1) .. "." .. toValue(payload, 3, 2)
+    serial_num = toValue(msg, 3, 2)
+    seq_num = string.byte(text, 3)
+
+    for k, v in pairs(resource_id) do
```

Comment:
I'd simply change the `id` field to the actual key for this map here, that way you don't have to iterate through them every time.

Reply:
That's a good idea. I revised it as your comment.